Fix handling of scoped define-funs

blishko opened the pull request:
This PR proposes to move the handling of defined functions from `Logic` to `Interpret` and treat them basically as macros when interpreting SMT-LIB2 file.
In the `Interpret`, we can tie their lifetime to the push frame where they were introduced and remove them on `pop`.

Fixes #512.

blishko commented on src/bin/Interpret.cc:
```diff
@@ -971,6 +972,13 @@ bool Interpret::defineFun(const ASTNode& n)
+bool Interpret::storeDefinedFun(std::string const & fname, const vec<PTRef> & args, SRef ret_sort, const PTRef tr) {
+    if (defined_functions.has(fname)) { return false; }
+
+    defined_functions.insert(fname, TemplateFunction(fname, args, ret_sort, tr), true); // TODO: ask config if defined are scoped
```
The standard introduced option `:global-declarations`, but I don't think we have it yet. We need to add it to the config.

aehyvari commented on src/bin/Interpret.h:
```diff
@@ -32,9 +32,36 @@ WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+    opensmt::ScopedVector<std::string> scopedNames;
+
+public:
+    bool has(const std::string& name) const { return defined_functions.find(name) != defined_functions.end(); }
+
+    void insert(const std::string& name, TemplateFunction && templ, bool scoped = false) {
```
Check for west-constness

blishko replied:
Fixed

aehyvari commented on src/bin/Interpret.h:
```diff
@@ -32,9 +32,36 @@ WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+        if (scoped) { scopedNames.push(name); }
+    }
+
+    TemplateFunction & operator[](const char* name) {
```
Check for west-constness

blishko replied:
Fixed

aehyvari reviewed (approved):
LGTM.  I left two small style comments.

blishko commented:
I fixed the style and added also the new option `:global-declarations`.
There is also a new test to exercise this option.
@aehyvari , do you want to check again?

aehyvari commented:
Ok I'll have a quick look!

aehyvari reviewed (approved):
Looks good!  For now we don't have the problem with the declarations being global but let's think about that later then...
